### PR TITLE
Make submit-queue run against other projects

### PR DESCRIPTION
**Pull request by foxish**
#1304

Added feature to access google-cloud-storage and specify bucket
and directory parameters from the command-line.

### CODE REVIEW

File: mungegithub/features/features.go [hunk 1]
```diff
@@ -92,3 +95,13 @@ func RegisterFeature(feat feature) error {
 	glog.Infof("Registered %#v at %s", feat, feat.Name())
 	return nil
 }
+
+// GetFeature returns a feature when requested by name.
+func GetFeature(name string) feature {
+	var feat feature
```
Comment (apelisse) [inline mungegithub/features/features.go]: No need to define variable first, just use `:=`

File: mungegithub/mungers/submit-queue.go [hunk 2]
```diff
@@ -381,8 +383,8 @@ func (sq *SubmitQueue) internalInitialize(config *github.Config, features *featu
 			gcs = utils.NewTestUtils("bucket", "logs", overrideUrl)
 		} else {
 			gcs = utils.NewWithPresubmitDetection(
-				utils.KubekinsBucket, utils.LogDir,
-				utils.PullKey, utils.PullLogDir,
+				sq.features.GCSInfo.BucketName, sq.features.GCSInfo.LogDir,
+				sq.features.GCSInfo.PullKey, sq.features.GCSInfo.PullLogDir,
```
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: I assume this is the  main part of the feature: we want to be able to change these value. Does it need to be a feature though?

Comment (foxish) [inline mungegithub/mungers/submit-queue.go]: I had a discussion with @lavalamp about this. GCSInfo needs to be used by two different mungers, and supplied from the commandline. Because it could be used by several mungers, it should exist outside any individual munger in my opinion. The options were:
* to make them global flags populated when mungegithub is run, in the main function (but that makes them sort of opaque).
* Introduce a new struct, which would in turn be passed into every munger in its Initialize() method (this needed changes to the Mungers interface).
* Making it a feature, similar to the feature 'gitrepos' used to read the locally fetched repository in blunderbuss.

I went with the third because it seemed like the most natural way to approach it, and its similarity to how 'gitrepos' works. If there is a better mechanism I may have missed, please let me know. :)

File: mungegithub/features/features.go (hunk 1 above)
Comment (foxish) [inline mungegithub/features/features.go]: Removing the 'GetFeature' function, as it is unused.

File: mungegithub/mungers/mungers.go [hunk 3]
```diff
@@ -69,15 +69,22 @@ func RequestedFeatures() []string {
 	return out.List()
 }
 
-// InitializeMungers will call munger.Initialize() for all mungers requested
```
Comment (apelisse) [inline mungegithub/mungers/mungers.go]: How is this related? What problem is it trying to fix?

File: mungegithub/mungers/submit-queue.go (hunk 2 above)
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: OK sgtm

File: mungegithub/mungers/mungers.go (hunk 3 above)
Comment (foxish) [inline mungegithub/mungers/mungers.go]: Previously, the mungers would first be validated & initialized, before the necessary features were initialized. The order was:

- mungers register & initialize
- features initialize

However, submit-queue.go would use the GCS feature during its initialization, so, it was necessary to initialize the features before the mungers. Hence, I broke it up such that the order becomes:

- mungers are registered
- list of features is obtained
- features initialize
- mungers initialize

Comment (apelisse) [inline mungegithub/mungers/mungers.go]: Cool, makes sense. Can you squash the three commits? Then LGTM

Comment (foxish) [inline mungegithub/mungers/mungers.go]: Done. Thank you for the quick review @apelisse.

File: mungegithub/features/features.go [hunk 4]
```diff
@@ -26,7 +26,8 @@ import (
 // Features are all features the code know about. Care should be taken
 // not to try to use a feature which isn't 'active'
 type Features struct {
-	Repos  *RepoInfo
+	Repos *RepoInfo
+	*GCSInfo
```
Comment (lavalamp) [inline mungegithub/features/features.go]: `GCSInfo *GCSInfo` unless you're purposely embedding the methods.

Comment (foxish) [inline mungegithub/features/features.go]: Did not know that it embedded methods that way. Fixed.

**Comment by lavalamp** [comment]
LGTM, thanks!

**Comment by lavalamp** [comment]
@eparis I'll merge this next time I push. You might want to take a look.

File: mungegithub/mungers/submit-queue.go (hunk 2 above)
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: I don't love the feature interface or how squirrely it is to use features correctly (I wrote it but someone smarter can probably rewrite it better) but I agree it seems like the right thing here...

File: mungegithub/mungegithub.go [hunk 5]
```diff
@@ -105,7 +105,7 @@ func main() {
 			}
 			errInit := mungers.InitializeMungers(&config.Config, &config.Features)
 			if errInit != nil {
```
Comment (lavalamp) [inline mungegithub/mungegithub.go]: Please use the pattern:

```
if err := mungers.InitializeMungers(&config.Config, &config.Features); err != nil {
```

Here and above on line 99. This way you can use the same variable name while avoiding this type of bug (err is only declared in the body of the if).

Comment (foxish) [inline mungegithub/mungegithub.go]: Done. Thanks! :)

**Comment by lavalamp** [comment]
Great, please squash your changes and we can merge this.

**Comment by foxish** [comment]
Done.
